linked_list_unit_test.py

```diff
@@ -29,7 +29,16 @@ def test_add_link_before(self):
     def test_add_link_before_out_of_bounds_empty_list(self):
         with self.assertRaises(IndexError):
             self.ll.add_link_before('Z', 26)
-
+    
+    # Tries to insert D at index 3, verifies an IndexError is raised. 
+    # (If list if not empty and have, say, N elements, valid indices for both methods are 0 ... N-1, inclusive.)
+    # CHECK the related piazza post: https://piazza.com/class/k8aktyvmw804ns?cid=141
+    def test_add_link_before_out_of_bounds_index_n(self):
+        self.ll.add_link_before("A", 0)
+        self.ll.add_link_before("B", 0)
+        self.ll.add_link_before("C", 1)
+        self.ll.add_link_before("D", 3) # -> Exception: Index out of bounds 

```

Comment:
This needs to be wrapped in a with self.assertRaises(IndexError) + try catch, in order to verify the operation actually raises the exception.

Reply:
The function test_add_link_before_out_of_bounds_index_n has already changed with another codes in order to verify IndexError or Exception and its message. Refers to "Files changed":
```
    # Tries to insert D at index 3, verifies an IndexError is raised. 
    # (If list if not empty and have, say, N elements, valid indices for both methods are 0 ... N-1, inclusive.)
    # CHECK the related piazza post: https://piazza.com/class/k8aktyvmw804ns?cid=141
    def test_add_link_before_out_of_bounds_index_n(self):
        self.ll.add_link_before("A", 0)
        self.ll.add_link_before("B", 0)
        self.ll.add_link_before("C", 1)

        try:
            with self.assertRaises(IndexError):
                self.ll.add_link_before("D", 3)
        except Exception as msg:
             self.assertEqual(msg.args[0], YOUR_EXCEPTION_MESSAGE)
```

Reply:
And the function test_add_link_before_fails_at_link_before_tail(self) is not the same as test_add_link_before_out_of_bounds_index_n(self). The former verifies the add_link_before function, but the latter verifies add_front function.

Reply:
After having dinner (in 2 hours? my husband just cooked so I can’t put off eating, sorry) I’m gonna fix another functions which requires verifications in order to check IndexError or Exception message simultaneously.

Reply:
Update: Fixed verification tests in order to check IndexError or Exception messages at the same time